Sandboxed containers operator to build secure execution enabled peer pod Image

ANJANA-A-R-K opened the pull request:
**- Description of the problem which is fixed/What is the use case**
Made changes to build secure execution enabled peer pod Image from Sandboxed containers operator

**- What I did**

**- How to verify it**

**- Description for the changelog**

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -22,3 +22,6 @@ data:
   BASE_OS_VERSION: "9.2"
+
+  # To Enable SE for IBM Z
+  SE_BOOT: "1"
```
Better to use `yes/no` or `true/false` if this is tied to changes in cloud-api-adaptor - https://github.com/confidential-containers/cloud-api-adaptor/commit/d5ca6f3d5997f0ae6112e030254a3776e4403f09

bpradipt commented on config/peerpods/peerpodssecret.yaml:
```diff
@@ -10,4 +10,4 @@ stringData:
   #REDHAT_OFFLINE_TOKEN: "" #Required to download rhel base image : Download token from https://access.redhat.com/management/api
-  
\ No newline at end of file
+  #HKD_CRT: "" #Download the certificate from https://www.ibm.com/support/resourcelink/api/content/public/host-key-documents.html
```
How about naming it to `HOST_CERTS`  or `HOST_KEY_CERTS` ?
It's ok to be little descriptive in the config param imho.

snir911 commented on config/peerpods/peerpodssecret.yaml:
```diff
@@ -10,4 +10,4 @@ stringData:
   #REDHAT_OFFLINE_TOKEN: "" #Required to download rhel base image : Download token from https://access.redhat.com/management/api
-  
\ No newline at end of file
+  #HKD_CRT: "" #Download the certificate from https://www.ibm.com/support/resourcelink/api/content/public/host-key-documents.html
```
The certificate is multi line, i'm not sure it will work as key's value in the CM, i think that in such cases usually best practice is to encode the file in base64 format and set as value or upload the file as standalone secret.
Having said that, can't we have the certificate embedded in the script rather than asking the user to set it? i.e. is this certificate frequently change ?

ANJANA-A-R-K replied:
Hi @snir911, The HKD is user dependent and differs for each user so the user should provide it, it is working with the current flow where we've provided the HKD as a value in peerpodssecret file and we've tested the same by deploying a worload.
```
[root@bastion-sl ~]# oc get pods -n default
NAME             READY   STATUS    RESTARTS   AGE
osc-caa-commit   1/1     Running   0          100s

```

savitrilh replied:
hi @snir911  HOSTKEY is for per z Server. This will vary for customer to customer. Hence this has to be provided as input. Hence we are adding as secret.

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -73,6 +72,19 @@ function create_libvirt_image() {
+        release=$(echo $BASE_OS_VERSION | awk -F "." '{ print $2 }')
+        if [ "$version" -lt 9 ]; then
+            error_exit "Libvirt Secure Execution supports RHEL OS version 9.4 or above"
+        else
```
nit: you can do it in the same conditional with **or**

savitrilh replied:
@snir911  Yes, we can achieve this with "or" but only concern is when "version" is 10 and release is less than 4 . This will break in future .

savitrilh replied:
@ANJANA-A-R-K  when version is 10 and release is less than 4. This condition will break.

ANJANA-ARK replied:
Modified the condition

snir911 reviewed (approved):
LGTM, added minor comment, thanks.

bpradipt commented on config/peerpods/peerpodssecret.yaml:
```diff
@@ -10,4 +10,4 @@ stringData:
   #REDHAT_OFFLINE_TOKEN: "" #Required to download rhel base image : Download token from https://access.redhat.com/management/api
-  
\ No newline at end of file
+  #HOST_KEY_CERTS: "" #Download the certificate from https://www.ibm.com/support/resourcelink/api/content/public/host-key-documents.html
```
@ANJANA-A-R-K  as I understand this is the cert content (as string) pasted in the secret. Have you tested it once to ensure that multi-line string is handled correctly ?

savitrilh replied:
@bpradipt  This has been tested by adding HOSTKEY in peer-pod-secret.yaml file.

@ANJANA-ARK Can you add all logs here.

ANJANA-ARK replied:
Yes we have tested it, we're providing the crt value like
```
stringData:
  CLOUD_PROVIDER: ""
  LIBVIRT_URI: ""
  LIBVIRT_POOL: ""
  LIBVIRT_VOL_NAME: ""
  REDHAT_OFFLINE_TOKEN: ""
  HOST_KEY_CERTS: |
    -----BEGIN CERTIFICATE-----
    .
    .
    .
    -----END CERTIFICATE-----
```
this is the packer log where HKD is picked up in the cloud api adaptor
```
==> qemu.rhel: Uploading qcow2/build-s390x-se-image.sh => ~/build-s390x-se-image.sh
==> qemu.rhel: Provisioning with shell script: /tmp/packer-shell3632283365
    qemu.rhel: Building SE podvm image for s390x
    qemu.rhel: Finding host key files
    qemu.rhel: found host key file: "/tmp/files/HKD.crt"

```
and we've tested by deploying a workload 
```
[root@bastion-sl ~]# oc get pods -n default
NAME             READY   STATUS    RESTARTS   AGE
osc-caa-commit   1/1     Running   0          100s

```

snir911 replied:
AFAIU user needs to make sure the certificate lines are aligned, otherwise, it may mess it, maybe worth to add a comment..

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -73,6 +72,15 @@ function create_libvirt_image() {
+    if [ "$SE_BOOT" == "true" ]; then
+        version=$(echo $BASE_OS_VERSION | awk -F "." '{ print $1 }')
+        release=$(echo $BASE_OS_VERSION | awk -F "." '{ print $2 }')
+        if [ "$version" -lt 9 ] || { [ "$version" -eq 9 ] && [ "$release" -lt 4 ]; }; then
```
```suggestion
        if [[ "$version" -lt 9 || ("$version" -eq 9 && "$release" -lt 4) ]]; then
```

bpradipt commented:
The commits needs to be squashed as well

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -73,6 +72,15 @@ function create_libvirt_image() {
+        version=$(echo $BASE_OS_VERSION | awk -F "." '{ print $1 }')
+        release=$(echo $BASE_OS_VERSION | awk -F "." '{ print $2 }')
+        if [[ "$version" -lt 9 || ("$version" -eq 9 && "$release" -lt 4) ]]; then
+            error_exit "Libvirt Secure Execution supports RHEL OS version 9.4 or above"
```
@ANJANA-A-R-K  if min base os version for SE is 9.4, shouldn't the default be changed accordingly here
https://github.com/openshift/sandboxed-containers-operator/blob/ff025162858eb5040d26cfda1e57520580983876/config/peerpods/podvm/libvirt-podvm-image-cm.yaml#L24

ANJANA-ARK replied:
I've changed it to 9.4 in libvirt-podvm-image-cm.yaml

ANJANA-A-R-K committed:
Secure execution (SE) enabled peepod for s390x

Sandboxed containers operator to build secure execution enabled peer pod Image

Signed-off-by: Anjana A R K <ANJANA.A.R.K1@ibm.com>

bpradipt reviewed (approved):
/lgtm
Thanks
